### ISSUE TITLE
Native Linux version

### ISSUE THREAD

**Issue by c-schmitz** [HeidiSQL/HeidiSQL#1482]
**Is your feature request related to a problem? Please describe.**
Because HeidiSQL is written in Delphi it should be possible to compile a native Linux version based on the same code base.
Has this been considered in the past?

**Describe the solution you'd like**
Compile a native version for Linux.

**Describe alternatives you've considered**
Running on wine is also fine.

**Comment by ansgarbecker** [comment]
This was considered many times already. Just because it's Delphi code does dot ensure it can be compiled for Linux.
HeidiSQL indeed uses the classic VCL components, which are only compilable for Windows programs. As far as I know you will need FireMonkey code in Delphi to compile a Linux build, and the newest Delphi environment.
This is not a simple task, but a separate project.

**Comment by gandalfthegreydev** [comment]
AWESOME

**Comment by ansgarbecker** [comment]
This is the first point where I am so far:

![Image](https://github.com/user-attachments/assets/fbc18892-a194-48bf-a1da-3942aeecbb7a)

It's incredibly easy for me to create a Lazarus/FPC project. On the other hand some important components available in Delphi don't have a replacement here, making a conversion nearly impossible:
* `TButtonedEdit`
* ~~`TComboBoxEx`~~ .. oh I just see there is one, on the Lazarus "Misc" palette
* `TSynUri` (well, not important)
* `TRegistry` (used for storing sessions and all settings) exists but is Windows-only
* some more minor units

That missing `TRegistry` could be replaced by storing settings in an XML file, probably, not sure. An .ini file does not provide folder structures which are required for session settings.

**Comment by c-schmitz** [comment]
I'd recommend storage in a  XML or JSON file, whatever is easier for your to apply. Imho JSON might be a bit easier to user.

**Comment by ansgarbecker** [comment]
Probably a dumb question, however: where do Linux apps normally store such JSON or XML setting files? I don't suppose in the application path?

**Comment by c-schmitz** [comment]
In the home directory of the user in a file starting with a . (dot) (so it is invisible by normal list commands)

So /home/someusername/.heidisql  would be  a proper location. If you have more than one file to store you could also create a subdirectory starting with a dot where you then store all the files.

**Comment by gandalfthegreydev** [comment]
> Probably a dumb question, however: where do Linux apps normally store such JSON or XML setting files? I don't suppose in the application path?

It depends on the distribution, but 99% in `~/.config/your-app` directory

https://specifications.freedesktop.org/basedir-spec/latest/

**Comment by gandalfthegreydev** [comment]
> In the home directory of the user in a file starting with a . (dot) (so it is invisible by normal list commands)
> 
> So /home/someusername/.heidisql would be a proper location. If you have more than one file to store you could also create a subdirectory starting with a dot where you then store all the files.

Better to use the `.config` directory, it's more or less standardized by every linux distro.

**Comment by gandalfthegreydev** [comment]
That's strange, usually, a standard C/CPP program find the library path during the configure phase when creating the make file.
I don't know Lazarus at all, probably you should ask to them to have better reply. Or you should try 3-4 common paths as fallback during the build time ? Start with /usr/lib then /usr/local/lib and so on....

**Comment by ansgarbecker** [comment]
So the [Free Pascal documentation](https://www.freepascal.org/docs-html/rtl/system/loadlibrary.html) recommends to specify absolute paths in `LoadLibrary()` - weird, isn't it? I mean, installing libmysqlclient is by intention done in the global /usr/lib/ structure or similar. I am even failing to download libmysqlclient.so - there are just too many client utilities and versions on mysql.com.

> No assumptions should be made about the location of the loaded library if a relative pathname is specified. The behaviour is dependent on the platform. Therefore it is best to specify an absolute pathname if possible.

**Comment by gandalfthegreydev** [comment]
Can't you try to load the library from different paths as long as you don't find the first available ?

**Comment by ansgarbecker** [comment]
I should probably search for with some wildcard in /usr/lib/ and /usr/local/lib/ - and provide the found absolute paths in the library dropdown of the session manager. So the user has control over the used one, and can try them.

**Comment by gandalfthegreydev** [comment]
yes, something similiar 

but million dollar question: on Windows the Lazarus build already works?

**Comment by ansgarbecker** [comment]
Icons, menus, and most controls on the main form are converted.
Library existence is checked per `ldconfig -p`
Database tree already shows databases and tables, filter boxes above also work
OS scaling works on Linux, in the screen set to 150%
Dark mode automatically overtaken from OS setting

![Image](https://github.com/user-attachments/assets/e6e86155-bdf5-4d75-80d5-dce3a427a855)

I will provide a first Linux binary soon, for the early birds.

**Comment by gandalfthegreydev** [comment]
Great step forward. Can't wait for the linux binary

**Comment by ansgarbecker** [comment]
Here's a first binary for Linux. Be aware it offers no session management and only some things on the main form work. But ut it already shows how possible things are with Lazarus and FreePascal.
Plus you need to connect via command line, like so:

    heidisql -h=localhost -u=root

[heidisql-linux-alpha.zip](https://github.com/user-attachments/files/19056675/heidisql-linux-alpha.zip)

**Comment by gandalfthegreydev** [comment]
i Will try

small suggestion, don't know if easy to add: in the very early stages, block all "write" queries

**Comment by gandalfthegreydev** [comment]
A lot of commits here .... Any new early adopter build to test ?

**Comment by ansgarbecker** [comment]
Yes, of course: take the attached zip, unzip it somewhere. But please note it's still highly incomplete. Pay attention, some things like the "_Drop database_" menu item already work as expected! Other things you can try out: grid export to clipboard or to file, the user manager, loading a csv file, reformat SQL, copy table and some more.

Important advices:
*  run per command line: `heidisql -h=localhost -u=root`. It's important to do per command line as the session manager is not yet functional. You will get a mostly empty main window when running without the parameters.
* be sure you have libmysql or libmariadb installed. I think I did that via `sudo apt install libmariadb-dev-compat`

[heidisql-12.10.1.1-ubuntulinux.zip](https://github.com/user-attachments/files/19208097/heidisql-12.10.1.1-ubuntulinux.zip)

![Image](https://github.com/user-attachments/assets/11baaa66-bd96-49d4-98a7-8530e831074f)

**Comment by gandalfthegreydev** [comment]
Probably it's expected, but I can see a lot of glitches, look at the attached image, seems different than yours. Yours is much more clean than mine

![Image](https://github.com/user-attachments/assets/dd1ed6cd-0cc5-4c39-9096-3cdead50316f)

**Comment by ansgarbecker** [comment]
You mean the two filter boxes on top of the database tree? These are edit boxes, just with too much height. Please compare the Windows version:

![Image](https://github.com/user-attachments/assets/58026691-75fe-40fa-9675-dba6e6696171)

**Comment by gandalfthegreydev** [comment]
I see. On mine looks like tabs, not like filters.

**Comment by rvveber** [comment]
This is really worth it, there are no tools as good as HeidiSQL on Linux.
I was looking forward to this day 🥳🚀

**Comment by gandalfthegreydev** [comment]
> This is really worth it, there are no tools as good as HeidiSQL on Linux. I was looking forward to this day 🥳🚀

Totally agree.
There is also another advantage: get rid of proprietary tools (like Delphi) in favor to an open source compiler (Lazarus)

**Comment by juergen852** [comment]
My first steps on Ubuntu 24.04.2 LTS

![Image](https://github.com/user-attachments/assets/3517f421-9ac0-47e3-ad87-d2e367ee5527)

**Comment by ansgarbecker** [comment]
I suppose you neither have libmysql nor libmariadb installed. You need to install at least one of them. I think I did that via  `sudo apt install libmariadb-dev-compat`

**Comment by evs-xsarus** [comment]
Is it possible to connect to another port? My db is running on 3380

**Comment by ansgarbecker** [comment]
Try installing libmariadb

**Comment by ansgarbecker** [comment]
Regarding the non-loadable plugins: the code still tries to load MySQL plugins from `/path/to/heidisql/plugins/`, which is problematic on Linux, as libmysql itself is also loaded from the system wide directory. According to [this post](https://stackoverflow.com/questions/49194719/authentication-plugin-caching-sha2-password-cannot-be-loaded) I found the `caching_sha2_password.so` plugin (and others) could be located in `/usr/lib/mariadb/plugin/` or in `/usr/local/mysql/lib/plugin/caching_sha2_password.so` - but how should HeidiSQL find out that location? Shall I hardcode it for Linux? (and which of them?)

**Comment by gandalfthegreydev** [comment]
`/usr/local/app-name` is used for custom-defined applications/libs installed by the user. They are not coming from the distro itself.
So, I think, if you find something in `/usr/local/` that should be preferred over the system-wide install, probably because the users installed something customized.

**Comment by gandalfthegreydev** [comment]
If you want to see the standard library dirs for the system, you should look at `/etc/ld.so.conf` and any files inside `/etc/ld.so.conf.d` PLUS `/usr/lib` and `/lib` (and their 64bit versions too: `/usr/lib64` `/lib64`

**Comment by gandalfthegreydev** [comment]
Try also to use `/usr/sbin/ldconfig -p` to see if the lib you need is listed.

**Comment by evs-xsarus** [comment]
@ansgarbecker I found this blog, maybe it helps: https://clarkkromenaker.com/post/library-dynamic-loading-linux/

**Comment by ansgarbecker** [comment]
The libraries are already detected via `/sbin/ldconfig -p` - I meant the _plugins_ used by libmysql and libmariadb, which are separated .so files and I think they are not listed by ldconfig.

**Comment by ansgarbecker** [comment]
I'm quite a bit proud to have the setting management and the session manager basically running. Especially the "Import settings" tool now works, with a textfile exported by the Windows version of HeidiSQL. Other things to note are:
* configuration is now managed in a settings.json file, in `~/.config/heidisql/`. The structure in it is identical to that in the Windows registry.
* libraries for SQLite and PostgreSQL are now detected (using the above mentioned `ldconfig` command)
* other usable dialogs: search/replace, user manager, CSV import, SQL reformatter, Export grid data
* edit boxes seem generally to high on Linux. This will be addressed soon
* other todos: designer forms for tables, views, procedures etc, preferences dialog, translation through .po files, query execution etc. -  still lots of things.

![Image](https://github.com/user-attachments/assets/8cfbc728-4ee9-4d15-8083-71c278814497)

Here's a second pre-alpha binary, zipped together with the ini files used by the SQL highlighter:

[heidisql-12.10.1.2-ubuntulinux.zip](https://github.com/user-attachments/files/19346592/heidisql-12.10.1.2-ubuntulinux.zip)

**Comment by wbob** [comment]
impressed by your progress, made my first sql steps in heidi many years ago.

Do you want feedback at this point or later?
* as far as I can see XDG doesn't spec private files / secrets, but it's good to "chmod 600" ~/.config/heidisql/settings.json - making it non-readable by world
* unix_socket type connection in mariadb (besides tcp/ip and named pipes)
* optional interoperability (import? checkbox?) with ~/.my.cnf (mysql/mariadb) or ~/.pgpass (postgres) - or even ~/.ssh config

**Comment by ansgarbecker** [comment]
Although in the meantime I learned many tweaks how to convert Delphi forms correctly for use in Lazarus, they display bad in Linux. Compiling them in Lazarus on Windows looks very fine and very similar to the official Delphi version, but in Ubuntu these forms get ridiculous huge edit boxes and comboboxes. Here's a comparison of the same dialog with the same code with the same default DPI setting in both OSes:

![Image](https://github.com/user-attachments/assets/051e1147-6e9a-4d40-95d0-91d1135c3b52)

And yes, I read that GTK themes may cause that. But how's a developer supposed to align controls so they display without overlapping? Note I'm not even talking about DPI issues - this is just all at 100% DPI.

**Comment by evs-xsarus** [comment]
Does this help? Taken from ChatGPT:

To compile forms in Lazarus that display correctly in GTK on Linux, follow these steps:

### Step 1: Install Lazarus and GTK
Ensure you have both **Lazarus** and **GTK** installed on your Linux system.

1. **Install Lazarus**:
   - On Ubuntu/Debian-based systems, you can install Lazarus with the following command:
     ```bash
     sudo apt-get install lazarus
     ```
   - On Fedora:
     ```bash
     sudo dnf install lazarus
     ```

2. **Install GTK (if not already installed)**:
   - For GTK3 (commonly used with modern Linux systems):
     ```bash
     sudo apt-get install libgtk-3-dev
     ```
   - For GTK2 (older applications):
     ```bash
     sudo apt-get install libgtk2.0-dev
     ```

### Step 2: Configure Lazarus to Use GTK
Make sure Lazarus is set to use the GTK widget set:

1. **Open Lazarus**.
2. Go to **Tools** -> **Options** (or press `Ctrl+Shift+O`).
3. Under **Environment**, go to **General**.
4. In the **Widgetset** dropdown, select either **GTK2** or **GTK3** depending on your preference and GTK version.
   - **GTK2** is still widely used for compatibility.
   - **GTK3** is the modern version, with better features and appearance.

5. **Click OK** to save your settings.

### Step 3: Create or Open Your Project
1. **Create a new form** or open an existing one in Lazarus.
2. Ensure that the form does not have any platform-specific components that may not work well with GTK (e.g., certain Windows-specific controls like TImageList). Stick to standard components for better compatibility.

### Step 4: Adjust Form and Component Properties for GTK
Some properties of the form and components might need to be tweaked to ensure they display correctly on Linux using GTK:

1. **Font**: GTK may display fonts differently, so it's a good idea to use a cross-platform font like `Arial`, `Sans`, or `Liberation Sans`.
   - Set the font for your form and components using the `Font` property.
   
2. **Alignments and Size**: GTK can sometimes behave differently than other widget sets (like Win32). Ensure components are properly aligned and that the form size is reasonable for GTK's default layout.
   - Test with different alignment settings (`alNone`, `alTop`, `alLeft`, etc.).

3. **Theme**: GTK uses different themes on different Linux environments, so ensure your form's look and feel are flexible:
   - Use **Borders** or **Margins** that will look good across various themes.
   - Avoid using custom controls that are not GTK-native (unless you are sure they are well-supported).

### Step 5: Compile and Test on Linux
After making adjustments, compile your application:

1. **Build the project** by selecting **Build** -> **Build Project** (or press `Shift+F9`).
2. **Run the project** by clicking on the green play button (or press `F9`).

Ensure you test the application on a variety of Linux distributions to check for display issues, especially with GTK versions.

### Step 6: Debug and Fix Display Issues
If the form does not display as expected, you can try the following:

1. **Check GTK versions**: Some distributions may have GTK2 or GTK3 libraries installed, which can affect how the form appears.
   - You may need to compile against a specific version of GTK (either GTK2 or GTK3) depending on what is installed on the target machine.
   
2. **Check for missing libraries**: If the form appears broken or incomplete, ensure that the necessary GTK libraries are installed on your system.

3. **Adjust layout**: If components are overlapping or not sized correctly, manually adjust the layout or use `Align` properties for components to ensure proper resizing.

By following these steps, you should be able to compile and run Lazarus applications with forms that display correctly using GTK on Linux.

**Comment by ansgarbecker** [comment]
Please avoid pasting long AI babble here, this ticket already has too much noise.

And no, that does not help. Using GTK3 instead of GTK2 in Lazarus compiles, but the executable crashes entirely. According to Lazarus, GTK3 support is in alpha state, which would explain a crash.
Setting Arial as font for input controls does not reduce their height, I guess the padding in these boxes is the core problem:

![Image](https://github.com/user-attachments/assets/de71406c-ac78-4e56-9933-970ad993c1d6)

**Comment by evs-xsarus** [comment]
Will do. Can you set another font for the linux builds? Arial isn't always installed, perhaps try 

![Image](https://github.com/user-attachments/assets/bae11e67-78ac-4542-a87f-f279efc08e42)

![Image](https://github.com/user-attachments/assets/f68db333-18b6-411d-ae5d-c475e5d761c0)

**Comment by gandalfthegreydev** [comment]
> Will do. Can you set another font for the linux builds? Arial isn't always installed, perhaps try

Arial is a Windows/Microsoft font (AFAIK), so we should assume that isn't available on Linux. Should be better to use an open source/free font, maybe shared together with the app package.

**Comment by ansgarbecker** [comment]
I have placed a first somehow usable alpha version on the [download page](https://www.heidisql.com/download.php), and wrote a short paragraph about it in the [forum](https://www.heidisql.com/forum.php?t=43993).

I will close this issue now, as this is heavily overloaded with commits already. Further things will go in separate tickets.

**Comment by c-schmitz** [comment]
Awesome - thank you!

**Comment by rvreugde** [comment]
Great! Thanks!

**Comment by gandalfthegreydev** [comment]
@ansgarbecker do you have a new pre-built linux realease to test ?

**Comment by ansgarbecker** [comment]
Not yet. After the last release I took care about some issues in the Windows version, and these changes will all merge into the Linux version as well. I am planning to release v12.11 soon, first the Windows version, and short afterwards for Linux. I am glad the basics are set up for the Linux version, as I was starting to get crazy with these unit and forms conversions. Hoping to get into normal waters soon, probably with a beta or even GA version.

**Comment by gandalfthegreydev** [comment]
Thank you.
Isn't easier for you to keep only one version, in example the Lazarus one, and build for all platforms ?

**Comment by ansgarbecker** [comment]
Yes, definitely. But that would mean to drop MS SQL and Interbase support as long as I don't have a solution for these. Also the user settings would need to be converted from his registry to the json file. Themes would be dropped permanently, as Lazarus has no support for these, which means no dark mode for the Windows users. Plus there are some minor missing drag/drop events which I could probably ignore.

**Comment by gandalfthegreydev** [comment]
I've seen the new linux release. There is no "beta"/"alpha" flag, it could be considered stable ?
AFAIK, ssh tunnes doesn't work (there is still the plink.exe/ssh.exe requirement, on linux that's useless as ssh is native and you don't have "ssh.exe" but only "ssh".

**Comment by gandalfthegreydev** [comment]
Correction: ssh tunnels works. Just set the real ssh path (ie `/usr/bin/ssh`) and change the ssh key with the public one and not the plink/putty version

**Comment by chosten** [comment]
SSH tunnel only works if I ssh the server from command line first to save the key fingerprint.
Any way to bypass that ?
